plugins/amdgpu/amdgpu_plugin.c

```diff
@@ -528,6 +529,116 @@ void amdgpu_plugin_fini(int stage, int ret)
 
 CR_PLUGIN_REGISTER("amdgpu_plugin", amdgpu_plugin_init, amdgpu_plugin_fini)
 
+struct thread_data {
+	pthread_t thread;
+	uint64_t num_of_bos;
+	uint32_t gpu_id;
+	pid_t pid;
+	struct kfd_criu_bo_buckets *bo_buckets;
+	BoEntriesTest **bo_info_test;
+	int ret;
+};
+
+void *dump_bo_contents(void *_thread_data)
+{
+        int i, ret = 0;
+        int num_bos = 0;
+        struct thread_data* thread_data = (struct thread_data*) _thread_data;
+        struct kfd_criu_bo_buckets *bo_buckets = thread_data->bo_buckets;
+        BoEntriesTest **bo_info_test = thread_data->bo_info_test;
+
+        pr_info("amdgpu_plugin: Thread[0x%x] started\n", thread_data->gpu_id);
+        for (i = 0; i < thread_data->num_of_bos; i++) {
+		if (bo_buckets[i].gpu_id != thread_data->gpu_id)
+                       continue;
+
+		num_bos++;
+		if (bo_buckets[i].bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_VRAM ||
+		    bo_buckets[i].bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_GTT) {
+			char *fname;
+			int mem_fd;
+
+			if (bo_info_test[i]->bo_alloc_flags & KFD_IOC_ALLOC_MEM_FLAGS_PUBLIC) {
+				int drm_fd;
```

Comment:
To reduce indentation further, you could move the two different dumping implementations (map BO or /proc/pid/mem) each into its own helper functions. But that's up to you.

Reply:
I like this idea as I think the functions in this file are generally too big, which in turn causes a lot of identation.

I will change it after as I think there was some additional optimizations that can be done to these loops:
1. For mmap section. we do not need to call open_drm_render_device for each BO. (can just mmap with different offsets)
2. For /proc/pid/mem section, do not need to fopen for each BO and can just call lseek with different offsets.

I am making some further changes to this function as part of the fix for tensorflow (protobuf cannot handle buffers > 4GB).

And I also started implemented multi-threading for the restore section so I will have more changes to this commit.